### ISSUE TITLE
Compatibility with MacOS Sonoma?

### ISSUE THREAD

**Issue by antoniopaolini** [quodlibet/quodlibet#4384]
Hello, 
I use the last MacOS version, 4.4.0 and I was happy until today.

Yesterday I upgraded the MacOS from Ventura to Sonoma and QuodLibet won't start anymore.

I don't have idea how to help you to understand better. 
I had installed also another player, Clementine, and it is not working too, but it show the window and close immediatly, and then start the Apple bug reporter (I don't know how it is called) but Qoud Libet don't start absolutely. 

Maybe there's a incompatibility with new MacOS?

Thanks.

**Comment by Richjn** [comment]
Confirming application crash after upgrade to Sonoma
I only use Ex-Falso but confirm that the applications segfaults with the following output when launched from a terminal.

`richard$ /Applications/ExFalso.app/Contents/MacOS/exfalso 
/Applications/ExFalso.app/Contents/MacOS/exfalso: line 65:  1197 Segmentation fault: 11  "$PYTHON" "$bundle_bin/$APP" "$@"
`

**Comment by antoniopaolini** [comment]
```
paolini$ ./QuodLibet.app/Contents/MacOS/quodlibet 
./QuodLibet.app/Contents/MacOS/quodlibet: line 65: 48776 Segmentation fault: 11  "$PYTHON" "$bundle_bin/$APP" "$@"
```

**Comment by jdykstra** [comment]
Full backtrace, memory map, etc. attached.

[Quod Libet on Sonoma Crash.txt](https://github.com/quodlibet/quodlibet/files/12777399/Quod.Libet.on.Sonoma.Crash.txt)

**Comment by jdykstra** [comment]
FWIW, QL 4.6.0 from the repository runs fine on Sonoma with dependencies supplied by the latest homebrew packages.

**Comment by Daoro** [comment]
> FWIW, QL 4.6.0 from the repository runs fine on Sonoma with dependencies supplied by the latest homebrew packages.
While we wait for a proper package, could you drop some quick hints how to do it?

**Comment by popeyebrown** [comment]
Thanks 
I downgraded back to Ventura.  I am  confused that there is seems to be three place to put QuodLibet.  Homebrew, MacPorts or the root.  I need to get it back to the root only.  

I’ll wait till you develop it in a couple weeks.

Thanks again

**Comment by antoniopaolini** [comment]
Hello, 
How the work is progressing? Could you give us some updates?

@popeyebrown downgrade the OS isn't a viable option for everybody.
Even though it is more simple to downgrade MacOS than find another player as good as Quod Libet. :-(

**Comment by popeyebrown** [comment]
I tried installing the dependance in Ventura however my knowledge with working with code is poor.  I tried to get it to work but no luck and frustrated and stopped.  I went back to my Linx OS on an another machine and it works just fine with Quodlibet-4-6-0.  

Thanks again!

**Comment by jdykstra** [comment]
Current status: Cannot build gettext-0.21.1 due to:

`libtool: link: /Applications/Xcode.app/Contents/Developer/usr/bin/gcc -dynamiclib  -o .libs/libtextstyle.0.dylib  .libs/gl_array_list.o .libs/basename-lgpl.o .libs/binary-io.o .libs/c-ctype.o .libs/c-strcasecmp.o .libs/c-strncasecmp.o .libs/cloexec.o .libs/concat-filename.o .libs/exitfail.o .libs/fatal-signal.o .libs/fd-hook.o .libs/fd-ostream.o .libs/file-ostream.o .libs/full-write.o .libs/get_ppid_of.o .libs/get_progname_of.o .libs/getprogname.o .libs/html-ostream.o .libs/html-styled-ostream.o .libs/ialloc.o .libs/iconv-ostream.o .libs/gl_list.o glthread/.libs/lock.o .libs/malloca.o .libs/math.o .libs/mem-hash-map.o .libs/memory-ostream.o .libs/noop-styled-ostream.o .libs/ostream.o .libs/printf-frexp.o .libs/printf-frexpl.o .libs/safe-read.o .libs/safe-write.o .libs/sig-handler.o .libs/sockets.o .libs/stat-time.o .libs/styled-ostream.o .libs/sys_socket.o .libs/term-ostream.o .libs/term-style-control.o .libs/term-styled-ostream.o glthread/.libs/threadlib.o .libs/unistd.o unistr/.libs/u8-mbtouc.o unistr/.libs/u8-mbtouc-aux.o .libs/xmalloc.o .libs/xalloc-die.o .libs/xconcat-filename.o .libs/xgethostname.o .libs/gl_xlist.o .libs/xsize.o .libs/xvasprintf.o .libs/xasprintf.o .libs/color.o .libs/misc.o .libs/version.o .libs/asnprintf.o .libs/asprintf.o .libs/error.o .libs/free.o .libs/obstack.o .libs/open.o .libs/printf-args.o .libs/printf-parse.o .libs/reallocarray.o .libs/snprintf.o .libs/stat.o .libs/strerror.o .libs/strerror-override.o .libs/vasnprintf.o .libs/vasprintf.o   -Wl,-force_load,./.libs/libcroco_rpl.a -Wl,-force_load,./.libs/libglib_rpl.a -Wl,-force_load,./.libs/libxml_rpl.a  -lncurses -L/Users/jwd/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/jhbuild_prefix/lib -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib -L/Users/jwd/.pyenv/versions/3.11.6/lib  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.13 -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.13 -Wl,-headerpad_max_install_names   -install_name  /Users/jwd/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/jhbuild_prefix/lib/libtextstyle.0.dylib -compatibility_version 2 -current_version 2.2 -Wl,-exported_symbols_list,.libs/libtextstyle-symbols.expsym ld: Undefined symbols: _iconv_ostream_create, referenced from: <initial-undefines> clang: error: linker command failed with exit code 1 (use -v to see invocation) make[5]: *** [libtextstyle.la] Error 1 make[4]: *** [all] Error 2 make[3]: *** [all-recursive] Error 1 make[2]: *** [all] Error 2 make[1]: *** [all-recursive] Error 1 make: *** [all] Error 2 *** Error during phase build of gettext: ########## Error running make -j 17  *** [9/16]`

iconv_ostream_create() is not being built because of autotools config option HAVE_ICONV. The reference is from a ld -u option created by libtool. Currently figuring out which is wrong. Remainder of QL build dependencies look fine.

**Comment by jdykstra** [comment]
https://lists.gnu.org/archive/html/bug-gettext/2023-10/msg00017.html

In parallel with working with the gettext maintainers, I'm using a patched version to continue work on Sonoma support.  It is currently unclear whether MacOS iconv regressed, and if so, what the impact on QL internationalization will be.

**Comment by jdykstra** [comment]
Apple switched to a buggy implementation of libiconv in MacOS 15.0.  https://lists.gnu.org/archive/html/bug-gettext/2023-10/msg00022.html.  

I can't assess the impact of the known bugs in this new implementation, so I am going to add the old (GNU) libiconv to the projects built by JHBuild.  

This change by Apple should be kept in mind if internationalization bugs are reported in QL running on MacOS 15.x.

**Comment by jdykstra** [comment]
meson error while building glib2:

../../../../jhbuild_checkoutroot/glib-2.74.7/meson.build:1:0: ERROR: Unable to detect linker for compiler `/Applications/Xcode.app/Contents/Developer/usr/bin/gcc -Wl,--version -L/Users/jwd/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/jhbuild_prefix/lib -L/Users/jwd/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/jhbuild_prefix/lib -arch x86_64 -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib -L/usr/local/opt/libiconv/lib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.13 -Wl,-headerpad_max_install_names -arch x86_64 -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.13 -I/Users/jwd/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/jhbuild_prefix/include -arch x86_64 -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -I/usr/local/opt/libiconv/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk`
stdout:
stderr: ld: unknown options: --version
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Possibly related:  https://github.com/mesonbuild/meson/issues/12159, https://github.com/mesonbuild/meson/issues/12399

**Comment by jdykstra** [comment]
AILED: gio/tests/defaultvalue.p/defaultvalue.c.o
/Applications/Xcode.app/Contents/Developer/usr/bin/gcc -Igio/tests/defaultvalue.p -Igio/tests -I../../../../jhbuild_checkoutroot/glib-2.74.7/gio/tests -I. -I../../../../jhbuild_checkoutroot/glib-2.74.7 -Iglib -I../../../../jhbuild_checkoutroot/glib-2.74.7/glib -Igmodule -I../../../../jhbuild_checkoutroot/glib-2.74.7/gmodule -Igobject -I../../../../jhbuild_checkoutroot/glib-2.74.7/gobject -Igio -I../../../../jhbuild_checkoutroot/glib-2.74.7/gio -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -I/Users/jwd/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/jhbuild_prefix/include -I/usr/local/opt/libiconv/include -fcolor-diagnostics -Wall -Winvalid-pch -Wextra -Wpedantic -std=gnu99 -O2 -g -D_GNU_SOURCE -fno-strict-aliasing -DG_DISABLE_CAST_CHECKS -Wimplicit-fallthrough -Wmisleading-indentation -Wunused -Wno-unused-parameter -Wno-cast-function-type -Wno-pedantic -Wno-format-zero-length -Wno-variadic-macros -Werror=format=2 -Werror=init-self -Werror=missing-include-dirs -Werror=pointer-arith -Werror=unused-result -Wstrict-prototypes -Wno-bad-function-cast -Werror=declaration-after-statement -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=pointer-sign -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=10.13 -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk '-DG_LOG_DOMAIN="GLib-GIO"' '-DGLIB_MKENUMS="gobject/glib-mkenums"' '-DGLIB_COMPILE_SCHEMAS="/Volumes/nobackup-cs/ql_bundle_build/quodlibet/dev-utils/osx_bundle/_home/.cache/jhbuild/build/glib-2.74.7/gio/glib-compile-schemas"' -UG_DISABLE_ASSERT -MD -MQ gio/tests/defaultvalue.p/defaultvalue.c.o -MF gio/tests/defaultvalue.p/defaultvalue.c.o.d -o gio/tests/defaultvalue.p/defaultvalue.c.o -c ../../../../jhbuild_checkoutroot/glib-2.74.7/gio/tests/defaultvalue.c
In file included from ../../../../jhbuild_checkoutroot/glib-2.74.7/gio/tests/defaultvalue.c:178:
gio/tests/giotypefuncs.inc:168:9: error: call to undeclared function 'g_osx_app_info_get_type'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
*tp++ = g_osx_app_info_get_type ();
        ^
gio/tests/giotypefuncs.inc:168:9: note: did you mean 'g_app_info_get_type'?
../../../../jhbuild_checkoutroot/glib-2.74.7/gio/gappinfo.h:150:13: note: 'g_app_info_get_type' declared here
GType       g_app_info_get_type                     (void) G_GNUC_CONST;
            ^
1 error generated.
[1000/1385] Linking target gio/tests/cxx-20
ld: warning: -undefined error is deprecated
[1013/1385] Compiling C object gio/tests/file.p/file.c.o
ninja: build stopped: subcommand failed.
*** Error during phase build of glib2: ########## Error running ninja   *** [24/88]


After several weeks of work, I have successfully built only 23 of the 88 QL dependencies.  There is a huge amount of bit rot in this bundle build.  I don't think we should plan on Sonoma support any time soon.

**Comment by jdykstra** [comment]
I now have 40 QL dependent packages building on Sonoma.  The total number of dependencies has grown to 90 packages.

**Comment by jdykstra** [comment]
All dependent packages are now building except for gst-plugins-bad, which is commented out.

**Comment by jdykstra** [comment]
Apple's switch to a buggy version of libiconv (see above) is causing many problems.  Many GNU packages have no provision for getting that library from any place other than the linker's default library directories.  Even when they do, JHBuild doesn't seem to respect package dependencies, and libiconv isn't built early enough.

**Comment by declension** [comment]
Great work @jdykstra. Is there a branch / CI / something you can link to?

**Comment by jdykstra** [comment]
> Great work @jdykstra. Is there a branch / CI / something you can link to?

The work-in-progress is on jdykstra/quodlibet, sonoma_support branch.  All packages built successfully when I was fixing problems as I went along, but a build from scratch fails.  Some package builds were order-dependent and in at least one case, the build picked up a library installed via Homebrew.

**Comment by jdykstra** [comment]
Capturing this so someone else doesn't go down the same rat hole.  GStreamer build aborting with:

`(<unknown>:26749): GLib-GObject-WARNING **: 00:34:24.883: cannot register existing type 'GstObject'
(<unknown>:26749): GLib-CRITICAL **: 00:34:24.884: g_once_init_leave: assertion 'result != 0' failed
(<unknown>:26749): GLib-GObject-CRITICAL **: 00:34:24.884: g_type_register_static: assertion 'parent_type > 0' failed
(<unknown>:26749): GLib-GObject-WARNING **: 00:34:24.884: cannot add private field to invalid (non-instantiatable) type '<invalid>'
(<unknown>:26749): GLib-CRITICAL **: 00:34:24.884: g_once_init_leave: assertion 'result != 0' failed
(<unknown>:26749): GLib-GObject-CRITICAL **: 00:34:24.884: g_type_register_static: assertion 'parent_type > 0' faile
(<unknown>:26749): GLib-GObject-WARNING **: 00:34:24.884: cannot add private field to invalid (non-instantiatable) type '<invalid>'
(<unknown>:26749): GLib-CRITICAL **: 00:34:24.884: g_once_init_leave: assertion 'result != 0' failed
Invalid GType function: 'gst_aggregator_pad_get_type'
Function 'gst_aggregator_pad_get_type' returned G_TYPE_INVALID
Command '['/Volumes/nobackup-cs/quodlibet/dev-utils/osx_bundle/_home/.cache/jhbuild/build/gstreamer-1.20.6/tmp-introspectp8hl5h_3/GstBase-1.0', '--introspect-dump=/Volumes/nobackup-cs/quodlibet/dev-utils/osx_bundle/_home/.cache/jhbuild/build/gstreamer-1.20.6/tmp-introspectp8hl5h_3/functions.txt,/Volumes/nobackup-cs/quodlibet/dev-utils/osx_bundle/_home/.cache/jhbuild/build/gstreamer-1.20.6/tmp-introspectp8hl5h_3/dump.xml']' returned non-zero exit status 1.`

Tried updating to gobject-introspection 1.78.0.  It requires glib2 >= 2.78.0.  Updated to glib2 2.78.0, which is the latest release.  Its build dies with:

`../../../../jhbuild_checkoutroot/glib-2.78.0/meson.build:2125:0: ERROR: Subproject exists but has no meson.build file` 

It appears the repo is broken, with directory subprojects/gvdb/ empty.

**Comment by jdykstra** [comment]
Subtle problem:  Most of the dependent packages were written on case-sensitive filesystems, so MacOS builds need to be done on a case-sensitive disk image volume.  But some of the GObject introspection utilities try to create a hard link from /var to a built artifact, which can't be done, and things go downhill from there.

**Comment by jdykstra** [comment]
The latest revision of JHBuild, 3.38.0, accepts the "supports-non-srcdir-build" module attribute as documented in its manual, but ignores it.  Out-of-tree builds of libtdb are badly broken.

**Comment by jdykstra** [comment]
I now have QuodLibet successfully building, and sort of executing on MacOS Sonoma.  Some libraries are looked for in the wrong place, and the QL window is a quarter of the size it's supposed to be.  But it's able to play FLAC's, and all of the UI functionality I've tried has worked.  I will continue testing and fixing problems as I find them.

This is still alpha-level code.   How do we find people to try it who will understand that it's not going to work perfectly and that we need their bug reports?  How do we distribute it?

The sonoma_support branch of github.com/jdykstra/quodlibet/ is up-to-date.

**Comment by jdykstra** [comment]
The first alpha release supporting Sonoma is at https://github.com/jdykstra/quodlibet/releases/tag/sonoma-alpha-1.  Please let me know if you install it.

**Comment by popeyebrown** [comment]
Thank you. Thank you.

Installed and works great.  It remembered my plugins and playlists, too!
Super job.

**Comment by popeyebrown** [comment]
Send me a paypal link and I'll donate to you.
thanks again
David Brown

**Comment by popeyebrown** [comment]
I am having trouble rediscovering the key command for highlighting a song:
the selected song e.g. bob seger? any ideas?  [image: Screenshot 2023-11-15
at 7.26.56 PM.png]

**Comment by jdykstra** [comment]
> I am having trouble rediscovering the key command for highlighting a song:
> the selected song e.g. bob seger? any ideas? 

Sorry, no.  I normally use QL via a touchscreen.

> Send me a paypal link and I'll donate to you.

Thanks, but unnecessary.  Donate to the GNU project instead.  https://www.gnu.org/help/help.html.

**Comment by spkm** [comment]
Works for me to. Thank you for your hard work @jdykstra!

I'll post any errors.

**Comment by alexander-e** [comment]
At first sight your alpha release seems to run just fine. Thanks a lot for the work! I'll be sure to report any bugs I may find.

**Comment by tgol** [comment]
Installed the alpha and tried a couple of features.
Playing music, tagging, renaming, replay gain...
No issues so far.
I missed this, thanks !

**Comment by jdykstra** [comment]
An unofficial 4.6.1 release of Quod Libet supporting MacOS Sonoma (and earlier releases) is at https://github.com/jdykstra/quodlibet/releases/tag/release-4.6.1.

@declension:  It's time to get this into your repo and website.  I assume the above DMG is suitable for putting on your downloads page.   For the pull request, do you want all 22 commits (some with not-so-great messages), or one squashed one?  I'll base it on main, but do you also want one based on your release-4.6.0 tag?

**Comment by declension** [comment]
Closed by #4433
